syntax = "proto2";

message Book {
  optional string ISBN = 1;
  optional string title = 2;
  optional string author = 3;

  enum Genre {
    FICTION = 0;
    NOVEL = 1;
    FANTASY = 2;
    POETRY = 3;
  }

  optional Genre genre = 4;
  optional int32 publish_year = 5;
}

message InventoryItem {
  optional string inventory_number = 1;
  oneof type {
    Book book = 2;
  }
  enum Status {
    AVAILABLE = 0;
    TAKEN = 1;
  }
  optional Status status = 3;
}

message HelloWorldRequest {
  optional string request = 1;
}

message HelloWorldResponse {
  optional string response = 1;
}

service HelloWorld {
  rpc HelloWorld(HelloWorldRequest) returns (HelloWorldResponse) {}
}

message GetBookByISBNRequest {
  optional string ISBN = 1;
}

message GetBookByISBNResponse {
  optional Book book = 1;
}

message CreateBookRequest {
  optional Book book = 1;
}

message CreateBookResponse {
  optional string response = 1;
}

service InventoryService {
  rpc GetBookByISBN(GetBookByISBNRequest) returns (GetBookByISBNResponse) {}
  rpc CreateBook(CreateBookRequest) returns (CreateBookResponse) {}
}